### ISSUE TITLE
Compatibility with Tensorflow Serving

### ISSUE THREAD

**Issue by r-wheeler** [google/sentencepiece#325]
Any idea how to best integrate the tensorflow op with tensorflow serving? 

Currently if this is used to train, when the tensorflow Graph is exported to a servable and ran with tensorflow serving a run time error will obviously occur. 

For example a model trained with this op trying to be loaded into tensorflow serving will result in: 

```
Loading servable: {name: xling } failed: Not Found: Op tyope not registered `SentencepieceEncodeSparse' in binary...
```

**Comment by arakawamoriyuki** [comment]
According to [this issues](https://github.com/tensorflow/serving/issues/678), it seems to be necessary to rebuild including [sentencepiece_processor_ops.so](https://github.com/google/sentencepiece/tree/master/tensorflow/tf_sentencepiece) in order to operate custom operation by serving.
We have confirmed that we can serve using the custom_op_paths parameter of [simple_tensorflow_serving](https://github.com/tobegit3hub/simple_tensorflow_serving#custom-op) as an alternative.
Also, if you are using Keras, it may work as long as it supports [KerasLayer](https://www.tensorflow.org/hub/api_docs/python/hub/KerasLayer) from tensorflow v2 to tf.hub.
I also want to know if there is a good solution that can be done now.

**Comment by taylorchu** [comment]
This is the ticket for adding sentencepiece to serving: https://github.com/tensorflow/serving/issues/1333

It is such a pain to recompile serving when an ops is added. After the plan to add sentencepiece to tf.contrib is scratched, we should find some other alternatives to support this in serving.

**Comment by stefanondisponibile** [comment]
@taylorchu did you have any chance (even painful) serving a model using custom `sentencepiece` ops **and** *TF Serving*? I've tried following [this](https://www.tensorflow.org/tfx/serving/custom_op#top_of_page) guide with no luck.

**Comment by r-wheeler** [comment]
@stefanondisponibile We have compiled tensorflow serving with sentencepiece. It very very complicated, This entire codebase had to be converted to build with bazel.

**Comment by stefanondisponibile** [comment]
Thanks for the feedback @r-wheeler .

**Comment by taylorchu** [comment]
@r-wheeler @stefanondisponibile yes, although at least mine doesn't sound as complicated as that. after sentencepiece and tf2 recent change, I need to patch it up again. It is always a lot of work without this merged into master.

**Comment by mhajiaghayi** [comment]
may I ask if it would be a compiled tensorflow that has the sentencepiece op? I've trained a model using sentencepiece and now in the run time, I can only call tensorflow in C++ environment and then load pb of my trained model. currently, i'm getting similar error in tf serving
_Not Found: Op tyope not registered `SentencepieceEncodeSparse' in binary_

what's my best option there?

**Comment by hbrylkowski** [comment]
@r-wheeler would it be possible to share your repository, or at least compiled tf-serving with sentencepiece?

**Comment by stefanondisponibile** [comment]
I've tried this in various flavors with no luck.

What I've done in particular is:

1. Cloned the [`serving`](https://github.com/tensorflow/serving) repo.
2. From the `serving` repo _root_ folder, I created a `tensorflow_serving/custom_ops` directory.
3. I cloned the the [`sentencepiece`](https://github.com/google/sentencepiece) repository inside the `custom_ops` directory.
4. I created a new `tensorflow_serving/custom_ops/sentencepiece/BUILD` file, which I filled like this:

```
licenses(["notice"])  # Apache 2.0

package(default_visibility = ["//visibility:public"])

cc_library(
    name ='sentencepiece',
    srcs = [
        "tensorflow/sentencepiece_processor_ops.cc",
    ],
    alwayslink = 1,
    deps = [
        "@org_tensorflow//tensorflow/core:framework",
    ]
)
```

5. I edited the `tensorflow_serving/model_servers/BUILD` file to add the custom op build target to `SUPPORTED_TENSORFLOW_OPS`, which now looks like this:

```
SUPPORTED_TENSORFLOW_OPS = if_v2([]) + if_not_v2([
    "@org_tensorflow//tensorflow/contrib:contrib_kernels",
    "@org_tensorflow//tensorflow/contrib:contrib_ops_op_lib",
    "//tensorflow_serving/custom_ops/sentencepiece:sentencepiece"
])
```

6. Ran `tools/run_in_docker.sh bazel build tensorflow_serving/model_servers:tensorflow_model_server` and crossed fingers (for a long, long, time), then received this error:

```
ERROR: /me/Documents/tf-sentencepiece/serving/tensorflow_serving/custom_ops/sentencepiece/BUILD:5:1: C++ compilation of rule '//tensorflow_serving/custom_ops/sentencepiece:sentencepiece' failed (Exit 1)
tensorflow_serving/custom_ops/sentencepiece/tensorflow/sentencepiece_processor_ops.cc:20:10: fatal error: sentencepiece_processor.h: No such file or directory
 #include "sentencepiece_processor.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Target //tensorflow_serving/model_servers:tensorflow_model_server failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 570.610s, Critical Path: 80.60s
INFO: 163 processes: 163 local.
FAILED: Build did NOT complete successfully
```

---

Of course I'm missing something, I wonder if anyone knows what 🤨

**Comment by alexpnt** [comment]
@stefanondisponibile you need to add the source headers to the build file and other core dependencies.

Something like this:

```
package(
    default_visibility = ["//visibility:public"],
    features = [
        "-layering_check",
        "-parse_headers",
    ],
)

licenses(["notice"])  # Apache 2, BSD, MIT

cc_library(
    name = "sentencepiece_processor",
    srcs = [
        "sentencepiece_processor_ops.cc",
    ],
    hdrs = [
        "sentencepiece_processor.h",
    ],
    alwayslink = 1,
    deps =
        [
            "@com_google_glog//:glog",
            "@org_tensorflow//tensorflow/core:framework",
            "@org_tensorflow//tensorflow/core:lib",
        ],
)


```

**Comment by stefanondisponibile** [comment]
thanks for the hint @alexpnt, but unfortunately that wasn't enough to compile sentencepiece with tensorflow serving successfully.

**Comment by alexpnt** [comment]
What error are you getting?

**Comment by stefanondisponibile** [comment]
@alexpnt basically it gives this error for various operations:

```
error: undefined reference to 'sentencepiece::xxx::yyy::zzz()'
```

It looks like all of those are referenced inside `sentencepiece_processor.h`, yet it fails with my BUILD configuration. 

This is the full stack:
```
ERROR: /me/Documents/tf-sentencepiece/serving/tensorflow_serving/model_servers/BUILD:372:1: Linking of rule '//tensorflow_serving/model_servers:tensorflow_model_server' failed (Exit 1)
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::ToTFStatus(sentencepiece::util::Status const&): error: undefined reference to 'sentencepiece::util::Status::error_message() const'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::ToTFStatus(sentencepiece::util::Status const&): error: undefined reference to 'sentencepiece::util::Status::code() const'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::Status()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::operator=(sentencepiece::util::Status const&)'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::~Status()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::~Status()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::SentencePieceProcessor::SentencePieceProcessor()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::operator=(sentencepiece::util::Status const&)'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::~Status()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::operator=(sentencepiece::util::Status const&)'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::~Status()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::SentencePieceProcessor::SentencePieceProcessor()'
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::SentencePieceBaseOp::SentencePieceBaseOp(tensorflow::OpKernelConstruction*): error: undefined reference to 'sentencepiece::util::Status::operator=(sentencepiece::util::Status const&)'
collect2: error: ld returned 1 exit status
Target //tensorflow_serving/model_servers:tensorflow_model_server failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 205.829s, Critical Path: 73.81s
INFO: 1 process: 1 local.
FAILED: Build did NOT complete 
```

**Comment by alexpnt** [comment]
Try to adapt the path to the full path on the #include lines. 

Something like: 

`tensorflow_serving/custom_ops/sentencepiece_processor/etc.h`

**Comment by stefanondisponibile** [comment]
Thank you for the feedback, I've tried this already too, in particular the following set up gives the same error.

On `sentencepiece_processor_ops.cc`:
`#include "tensorflow_serving/custom_ops/sentencepiece_processor/sentencepiece_processor.h"`

`BUILD` file:
```
package(
    default_visibility = ["//visibility:public"],
    features = [
        "-layering_check",
        "-parse_headers",
    ],
)

licenses(["notice"])  # Apache 2, BSD, MIT

cc_library(
    name = "sentencepiece_processor",
    srcs = [
        "sentencepiece_processor_ops.cc",
    ],
    hdrs = [
      "//tensorflow_serving/custom_ops/sentencepiece_processor:sentencepiece_processor.h",
    ],
    alwayslink = 1,
    deps = [
        # "@com_google_glog//:glog",
        "@org_tensorflow//tensorflow/core:framework",
        "@org_tensorflow//tensorflow/core:lib",
    ],
)
```

**Error:**
```
ERROR: /me/Documents/tf-sentencepiece/serving/tensorflow_serving/model_servers/BUILD:372:1: Linking of rule '//tensorflow_serving/model_servers:tensorflow_model_server' failed (Exit 1)
bazel-out/k8-opt/bin/tensorflow_serving/custom_ops/sentencepiece_processor/_objs/sentencepiece_processor/sentencepiece_processor_ops.o:sentencepiece_processor_ops.cc:function sentencepiece::(anonymous namespace)::ToTFStatus(sentencepiece::util::Status const&): error: undefined reference to 'sentencepiece::util::Status::error_message() const'
[...]
```

**Comment by sooneecheah** [comment]
Hey @stefanondisponibile , not sure if you are still working on this, but we recently got this working. The steps to get it working with tensor serving v 1.14 were:

- Compile protobuf 3.7.0 from sources (https://github.com/protocolbuffers/protobuf/tree/master/src for instructions), but instead of just running `.configure`, run `./configure CXXFLAGS="-D_GLIBCXX_USE_CXX11_ABI=0" ` before the `make` and `make install`.
- Inside `custom_ops/sentencepiece_processor`, clone the entire `sentencepiece` repo (i.e. this repo) in.
- Compile the cloned SentencePiece from sources, but instead of just running `cmake ..`, run `cmake -DSPM_USE_BUILTIN_PROTOBUF=OFF -DSPM_ENABLE_TENSORFLOW_SHARED=ON ..`

The `custom_ops/sentencepiece_processor` folder should have a BUILD file that looks like:
```
package(
    default_visibility = ["//visibility:public"],
    features = [
        "-layering_check",
        "-parse_headers",
    ],
)

licenses(["notice"])  # Apache 2, BSD, MIT

cc_library(
    name = "sentencepiece_processor_ops",
    srcs = [
        "sentencepiece/tensorflow/sentencepiece_processor_ops.cc"
    ],
    hdrs = [
        "sentencepiece/src/sentencepiece_processor.h"
    ],
    strip_include_prefix = "sentencepiece/src/",
    alwayslink = 1,
    deps =
        [
            "@org_tensorflow//tensorflow/core:framework",
            "@org_tensorflow//tensorflow/core:lib",
            "@com_google_protobuf//:protobuf_headers",
            "@com_google_protobuf//:protobuf",
            ":sentencepiece"
        ]
)

cc_library(
    name = "sentencepiece",
    srcs = [
        "sentencepiece/build/src/libsentencepiece.a",
        "sentencepiece/build/src/libsentencepiece_train.a"
    ],
    hdrs = glob(
        ["sentencepiece/src/*.h"]
    ),
    deps = [
        "@com_google_protobuf//:protobuf_headers",
        "@com_google_protobuf//:protobuf",
    ],
    strip_include_prefix = "sentencepiece/src/"
)
```

- Your `tensorflow_serving/model_servers/BUILD` looks fine
- Compile Tensor Serving as per normal

Hope this helps, let me know if you run into any issues.

**Comment by stefanondisponibile** [comment]
Hey @sooneecheah, thank you for the detailed feedback! I actually gave up and switched to _simple_tensorflow_serving_ on my project, but I'm very interested in this and will try your build steps soon.

**Comment by stefanondisponibile** [comment]
I can confirm @sooneecheah build steps worked for me too 🎉
For anyone interested or in need of this, I used [this](https://gist.github.com/stefanondisponibile/a3d9388b146db82a70fb80b391b6e3c7) bash script to follow the steps above.

**Comment by davislf2** [comment]
Hi @stefanondisponibile, I followed your code and changed some parts to adapt to CentOS. However, I got errors in the last step:

```
CWD=$(pwd)
GPROTOBUF_URL=https://github.com/protocolbuffers/protobuf/releases/download/v3.7.0/protobuf-cpp-3.7.0.tar.gz

compile_gprotobuf () {
    sudo yum install autoconf automake libtool curl make gcc-c++ unzip && # gcc-c++ instead of (Ubuntu) g++
    wget $GPROTOBUF_URL &&
    fname=protobuf-cpp-3.7.0.tar.gz &&
    tar xvzf $fname && rm $fname &&
    cd protobuf-3.7.0 &&
    ./configure CXXFLAGS="-D_GLIBCXX_USE_CXX11_ABI=0" &&
    make &&
    make check &&
    sudo make install &&
    sudo ldconfig # refresh shared library cache. 
    cd $CWD
}

build_tensorflow_serving_and_sentencepiece () {
    git clone -b 'r1.14' --single-branch --depth 1 https://github.com/tensorflow/serving.git &&
    mkdir -p serving/tensorflow_serving/custom_ops/sentencepiece_processor &&
    git clone https://github.com/google/sentencepiece.git serving/tensorflow_serving/custom_ops/sentencepiece_processor/sentencepiece &&
    sudo yum install gcc gcc-c++ kernel-devel && # instead of (Ubuntu) sudo apt-get install build-essential
    sudo yum install cmake && 
    sudo yum provides */pkg-config &&
    sudo yum install google-perftools google-perftools-devel && # (Ubuntu) sudo apt-get install libgoogle-perftools-dev &&
    cd serving/tensorflow_serving/custom_ops/sentencepiece_processor/sentencepiece &&
    mkdir build &&
    cd build &&

    cmake -DSPM_USE_BUILTIN_PROTOBUF=OFF -DSPM_ENABLE_TENSORFLOW_SHARED=ON .. &&
    make -j $(nproc) &&
    sudo make install &&
    sudo ldconfig -v &&

    cd $CWD &&
    # cp ./build ./serving/tensorflow_serving/custom_ops/sentencepiece_processor/build && # I can't find this BUILD directory
    sed -i.bak '/@org_tensorflow\/\/tensorflow\/contrib:contrib_ops_op_lib/a\    "\/\/tensorflow_serving\/custom_ops\/sentencepiece_processor:sentencepiece_processor_ops",' ./serving/tensorflow_serving/model_servers/BUILD &&
    sed -i '/name = "tensorflow_model_server",/a\    linkopts = ["-Wl,--allow-multiple-definition", "-Wl,-rpath,/usr/lib"],' ./serving/tensorflow_serving/model_servers/BUILD
    
    wget https://copr.fedorainfracloud.org/coprs/vbatts/bazel/repo/epel-7/vbatts-bazel-epel-7.repo
    sudo cp vbatts-bazel-epel-7.repo /etc/yum.repos.d/vbatts-bazel-epel-7.repo
    sudo yum install bazel
    # https://docs.bazel.build/versions/master/install-redhat.html

    cd serving && 
    
    tools/run_in_docker.sh bazel build tensorflow_serving/model_servers:tensorflow_model_server
```

I removed `--max_idle_secs` because it gives me an error `ERROR: Unrecognized option: --max_idle_secs=60`. But this one will also cause a problem in `Fetching @org_tensorflow`, which shows
```
ERROR: error loading package '': in /my_path/serving/tensorflow_serving/workspace.bzl: Encountered error while reading extension file 'tensorflow/workspace.bzl': no such package '@org_tensorflow//tensorflow': java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/tensorflow/tensorflow/archive/87989f69597d6b2d60de8f112e1e3cea23be7298.tar.gz, https://github.com/tensorflow/tensorflow/archive/87989f69597d6b2d60de8f112e1e3cea23be7298.tar.gz] to /my_path/serving/.cache/_bazel_opc/3fa7ae51721c1323b37adf18f7a53821/external/org_tensorflow/87989f69597d6b2d60de8f112e1e3cea23be7298.tar.gz: All mirrors are down: []
```

I tried several methods to change the http_proxy (Use `HTTP_PROXY`, download the tar.gz file and put it in the dir, `--action_env=HTTP_PROXY=$HTTP_PROXY`, ...), but all of them failed.

Finally, I changed the image from `IMAGE="tensorflow/serving:nightly-devel"` to `IMAGE="tensorflow/serving:1.14.0"`, and it doesn't require bazel fetch. However, it shows
`unknown argument: bash`, and also `FileSystemStoragePathSource encountered a filesystem access error: Could not find base path /models/model for servable model`. Do you have any idea how I can assign the correct model base? Thanks.

**Comment by stefanondisponibile** [comment]
Hi @davislf2, are you sure this wasn't simply a connection problem 🤔 Moreover I think the second error, telling you "could not find base path..." isn't related to the compilation, it's just TensorFlow serving running without finding a servable at that path.

**Comment by davislf2** [comment]
Hi @stefanondisponibile, thank you for responding! I have tried wget with my http_proxy like 
`wget https://github.com/tensorflow/tensorflow/archive/87989f69597d6b2d60de8f112e1e3cea23be7298.tar.gz -e use_proxy=yes -e http_proxy="http://my_proxy:80"` or `HTTP_PROXY="http://my_proxy:80" HTTPS_PROXY="http://my_proxy:80"` then `wget`, and I can download the file in either way. However, I still can't download through bazel. Would you think there might be other connection issues I neglected?

Also, I am wondering what kind of file I would expect to see before I start `run_in_docker.sh`? As I know, if I use a standard way to do TF-serving, then I will need a SavedModel (.pb and variables) and run the following command:
```
docker run -p 8501:8501 \
    --mount type=bind,source=$path_model,target=/models/$name_model \
    -e MODEL_NAME=$name_model -t tensorflow/serving
```

However, what I feel weird is that until the last line of `run_in_docker.sh`, I only see this line to serve the model: `docker run ${RUN_OPTS[@]} ${IMAGE} bash -c "$(get_switch_user_cmd) ${CMD}"`. I am not sure where I can include the model (e.g. https://tfhub.dev/google/universal-sentence-encoder-multilingual-large/1). Do you have any idea? Thanks.

**Comment by stefanondisponibile** [comment]
I'm sorry, but I'm not sure I can help with that connection problem. If you need the compiled version I can pass it somehow or (maybe better) dockerize it and therefore make it available via DockerHub. However, I opened an issue on Tensorflow Serving (#1490) and as you can see hopefully we will no longer need to compile tf serving from scratch.
The problem with the fact that tf serving isn't finding your model is a bit out of topic for this issue, but if you read carefully through the ‘run_in_docker.sh‘ file, you will see two volumes are declared for the running container, one in your working directory and another at ‘tmp‘, you should put your model in any path reachable from the container (so in your case this would be something like ‘/tmp/yourmodel’ or ‘$(pwd)/models/yourmodel‘).

**Comment by davislf2** [comment]
Thank you, @stefanondisponibile. I would really appreciate if you can share the docker via DockerHub. That would be a great help for me to replicate your works!

Sorry, the serving model is a bit off the topic. My understanding is that after we run `tools/run_in_docker.sh bazel build tensorflow_serving/model_servers:tensorflow_model_server` to build the tf-serving base docker image, and then we can serve it with our model [through this command from the doc](https://www.tensorflow.org/tfx/serving/custom_op#serve_a_model_containing_your_custom_op):
```
tools/run_in_docker.sh -o "-p $port_model:8501" \
bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server \
--rest_api_port=$port_model --model_name=$my_model --model_base_path=$my_model_path
```
Do I understand it correctly? However, this command shows me this error:
`# sh: 1: bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server: not found` It looks quite weird for me.

**Comment by int64max** [comment]
> 
> 
> I can confirm @sooneecheah build steps worked for me too 🎉
> For anyone interested or in need of this, I used [this](https://gist.github.com/stefanondisponibile/a3d9388b146db82a70fb80b391b6e3c7) bash script to follow the steps above.

Thanks all who've been helpful on this thread. I'm in the same conundrum. I.e., I want to use multilingual-2, and I'm unable to.

I followed the steps included here. I even ran the bash script linked by @stefanondisponibile (thanks btw!). Yet, I still get the following error [1]. I've used 'alwayslink=1' in the bazel rule for the 'sentencepiece_processor_ops' library. And I've added the 'sentencepiece_processor_ops' rule to 'SUPPORTED_TENSORFLOW_OPS' in tensorflow_serving/model_servers/BUILD. All the builds succeeded from the bash script. And I'm able to even start the server via "tools/run_in_docker.sh bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server". Yet, I still get the op type not registered error when I query the server.

I've run out of ideas. I'll highly appreciate any pointers. Thanks.

[1]
2019-11-19 07:53:36.117011: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1502] OP_REQUIRES failed at partitioned_function_ops.cc:112 : Not found: Op type not registered 'SentencepieceOp' in binary running on <>. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.

**Comment by sooneecheah** [comment]
Hey @int64max - I haven't tested this out but I wonder if the issue with multilingual-2 (I assume you mean this: https://tfhub.dev/google/universal-sentence-encoder-multilingual-large/2) is that it is trained in TF2.0 - the steps I had above builds Tensor Serving 1.14. I'm not sure if something has changed in 2.0 that is causing the error. Does multilingual-1 load with the binary you've built?

**Comment by int64max** [comment]
Thanks for the reply. Sorry for the confusion. By multilingual-2, I mean  https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/2.

With multilingual-qa-1, when I start the server, it gives me the following error [1]. I downloaded the model from https://tfhub.dev/google/universal-sentence-encoder-multilingual-qa/1?tf-hub-format=compressed, extracted it into /tmp/multilingual-qa-1/1, and then ran:

tools/run_in_docker.sh -o "-p 8501:8501" bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server --rest_api_port=8501 --model_name=test --model_base_path=/tmp/multilingual-qa-1/

Any ideas? Thanks.

[1]
2019-11-19 08:07:14.407251: E tensorflow_serving/util/retrier.cc:37] Loading servable: {name: test version: 1} failed: Not found: Could not find meta graph def matching supplied tags: { serve }. To inspect available tag-sets in the SavedModel, please use the SavedModel CLI: `saved_model_cli`

**Comment by davislf2** [comment]
Hi @sooneecheah, I tried [multilingual-large-1](https://tfhub.dev/google/universal-sentence-encoder-multilingual-large/1) but I didn't get good luck. 

I changed the last line of `tools/run_in_docker.sh` to `docker run ${RUN_OPTS[@]} ${IMAGE} bash`, and run `bazel build tensorflow_serving/model_servers:tensorflow_model_server` in the container. 

After successful build, I saw the `bazel-bin/` directory in the container, and I ran `bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server \ --rest_api_port=8501 --model_name=use_mul_l --model_base_path=/tmp/use_mul_l` in the container, but I still got the error of `Op type not registered 'SentencepieceEncoderSpare'` [1].

Other than that, I also did `docker commit b5c57bc8d18d sentencepiece-image` and ran `tools/run_in_docker.sh bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server \ --rest_api_port=8501 --model_name=use_mul_l --model_base_path=/tmp/use_mul_l` with the last line of `tools/run_in_docker.sh` been replaced with `docker run ${RUN_OPTS[@]} sentencepiece-image:latest bash`. I saw `bazel-bin\` was still in the container, but I still get the same error [1] while I ran `bazel-bin/tensorflow_....` as above. 

Do you know how could I check whether 'SentencepieceEncoderSpare' has been correctly compiled by bazel? Or is there any other way to run and serve this image? Thanks.

> [1] 
> 2019-11-20 01:08:10.161783: I external/org_tensorflow/tensorflow/cc/saved_model/loader.cc:333] SavedModel load for tags { serve }; Status: fail: Not found: Op type not registered 'SentencepieceEncodeSparse' in binary running on my-machine. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.. Took 650587 microseconds.
> 2019-11-20 01:08:10.161841: E tensorflow_serving/util/retrier.cc:37] Loading servable: {name: use_mul_l version: 1} failed: Not found: Op type not registered 'SentencepieceEncodeSparse' in binary running on my-machine. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.

**Comment by stefanondisponibile** [comment]
Sorry @davislf2 and @int64max but I don't have any clue of why @sooneecheah steps aren't working for you.

@davislf2 that `Op type not registered 'SentencepieceEncoderSparse` error it really sounds like it really literally did **not** registered that operation. You could maybe try starting from scratch without using my bash script and going carefully through @sooneecheah steps on your own.

Operations' names change from `use-mul-l:1` and `use-mul-l:1`, but this shouldn't be an issue, all that matters are references in the `BUILD` files.

If this is an option for you I'd consider using SimpleTensorFlow Serving at this point.
I don't know if that could help but if you want you can run [this](https://hub.docker.com/repository/docker/stefanondisponibile/simple-tf-use) docker container to have Simple TensorFlow Serving with `use-mul-l:1` built in to use it "as a service" (simply `POST` `{"model_name": "default", "model_version": "001", "data": {"text": ["foo"]}` to `localhost:8501/`) .

Sorry but I didn't have time yet to see how to dockerize my compiled TF Serving version. 

I actually hope TensorFlow Serving will soon include all current `tensorflow_text` operations which now hosts `sentencepiece` ops, things will be easier at that point.

**Comment by davislf2** [comment]
Hi @stefanondisponibile , I finally figured out why I can't get the Op. Based on your script, [gobrewers14](https://stackoverflow.com/users/3222797/gobrewers14) provides another solution to [wrap all works you have done in the docker image](https://stackoverflow.com/questions/58925650/build-file-from-cmake-for-customized-op-tensorflow-serving). @int64max You can try this approach to see whether it works for you. To sum up, here is the progress I did to make it works:

1. Use default tf-serving image to serve the model you want to run, and you shall see the error message like this:

   ```sh
   Op type not registered 'SentencepieceEncodeSparse' in binary running on my-machine.
   ```

   Keep the name of the Op you need to registered, and this name will be used later.

   Note: Different models will name their sentencepiece Op differently. e.g. [universal sentence encoder multilingual large/1](https://tfhub.dev/google/universal-sentence-encoder-multilingual-large/1) uses `SentencepieceEncodeSparse`, but [universal sentence encoder multilingual large/2](https://tfhub.dev/google/universal-sentence-encoder-multilingual-large/2) uses `SentencepieceOp`.

2. Create a file `sentencepiece.Dockerfile` like the following script, and
   - Modify or remove proxy setting as you need
   - Replace `SentencepieceEncodeSparse` to the Op name you get from step 1

```py
FROM ubuntu:16.04 as base_build

ENV http_proxy 'http://your-proxy:port'
ENV https_proxy 'http://your-proxy:port'

ARG TF_SERVING_BRANCH=r1.14
ARG TF_SERVING_COMMIT=head

RUN apt-get update \
  && apt-get install -y --no-install-recommends \
    autoconf \
    automake \
    build-essential \
    ca-certificates \
    cmake \
    curl \
    git \
    libcurl3-dev \
    libfreetype6-dev \
    libgoogle-perftools-dev \
    libpng-dev \
    libtool \
    libzmq3-dev \
    make \
    mlocate \
    openjdk-8-jdk\
    openjdk-8-jre-headless \
    pkg-config \
    python3-dev \
    software-properties-common \
    swig \
    unzip \
    wget \
    zip \
    zlib1g-dev \
    gdb \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*

ARG PROTOBUF_VERSION=3.7.0
RUN wget https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOBUF_VERSION}/protobuf-cpp-${PROTOBUF_VERSION}.tar.gz \
  && tar -xzvf protobuf-cpp-${PROTOBUF_VERSION}.tar.gz \
  && cd protobuf-${PROTOBUF_VERSION} \
  && ./configure CXXFLAGS="-D_GLIBCXX_USE_CXX11_ABI=0" \
  && make -j "$(nproc)" \
  && make install \
  && ldconfig

RUN curl -fSsL -O https://bootstrap.pypa.io/get-pip.py \
  && python3 get-pip.py \
  && rm get-pip.py

RUN pip3 --no-cache-dir install \
  future>=0.17.1 \
  grpcio \
  h5py \
  keras_applications>=1.0.8 \
  keras_preprocessing>=1.1.0 \
  mock \
  numpy \
  requests

RUN set -ex \
  && ln -s /usr/bin/python3 usr/bin/python \
  && ln -s /usr/bin/pip3 /usr/bin/pip

ENV BAZEL_VERSION 0.24.1
WORKDIR /
RUN mkdir /bazel \
  && cd /bazel \
  && curl \
    -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" \
    -fSsL -O https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh \
  && curl \
    -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" \
    -fSsL -o /bazel/LICENSE.txt https://raw.githubusercontent.com/bazelbuild/bazel/master/LICENSE \
  && chmod +x bazel-*.sh \
  && ./bazel-$BAZEL_VERSION-installer-linux-x86_64.sh \
  && cd / \
  && rm -f /bazel/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh

WORKDIR /tensorflow-serving
RUN git clone --branch=${TF_SERVING_BRANCH} https://github.com/tensorflow/serving . \
  && git remote add upstream https://github.com/tensorflow/serving.git \
  && if [ "${TF_SERVING_COMMIT}" != "head" ]; then git checkout ${TF_SERVING_COMMIT} ; fi

FROM base_build as binary_build

# clone sentencepiece repo and build
RUN set -ex \
  && mkdir -p tensorflow_serving/custom_ops/sentencepiece_processor \
  && working_dir=`pwd` \
  && git clone https://github.com/google/sentencepiece.git tensorflow_serving/custom_ops/sentencepiece_processor/sentencepiece \
  && cd tensorflow_serving/custom_ops/sentencepiece_processor/sentencepiece \
  && mkdir build \
  && cd build \
  && cmake -DSPM_USE_BUILTIN_PROTOBUF=OFF -DSPM_ENABLE_TENSORFLOW_SHARED=ON .. \
  && make -j $(nproc) \
  && make install \
  && ldconfig

COPY BUILD ./BUILD
RUN cp BUILD tensorflow_serving/custom_ops/sentencepiece_processor/BUILD \
  && sed -i.bak '/@org_tensorflow\/\/tensorflow\/contrib:contrib_ops_op_lib/a\ "\/\/tensorflow_serving\/custom_ops\/sentencepiece_processor:SentencepieceEncodeSparse",' \
       tensorflow_serving/model_servers/BUILD \
  && sed -i '/name = "tensorflow_model_server",/a\    linkopts = ["-Wl,--allow-multiple-definition", "-Wl,-rpath,/usr/lib"],' \
       tensorflow_serving/model_servers/BUILD

# Build, and install TensorFlow Serving
ARG BUILD_OPTIONS="--config=nativeopt"
ARG BAZEL_OPTIONS=""
# build bazel
RUN bazel build \
      --color=yes \
      --curses=yes \
      ${BAZEL_OPTIONS} \
      --verbose_failures \
      --output_filter=DONT_MATCH_ANYTHING \
      ${BUILD_OPTIONS} \
      tensorflow_serving/model_servers:tensorflow_model_server \
  && cp bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server /usr/local/bin/

CMD ["/bin/bash"]

```

3. Create `BUILD` file and change the name `SentencepieceEncodeSparse` to the Op name you get from step 1 as well.

```sh
package(
    default_visibility = ["//visibility:public"],
    features = [
        "-layering_check",
        "-parse_headers",
    ],
)

licenses(["notice"])  # Apache 2, BSD, MIT

cc_library(
    name = "SentencepieceEncodeSparse",
    srcs = [
        "sentencepiece/tensorflow/sentencepiece_processor_ops.cc"
    ],
    hdrs = [
        "sentencepiece/src/sentencepiece_processor.h"
    ],
    strip_include_prefix = "sentencepiece/src/",
    alwayslink = 1,
    deps =
        [
            "@org_tensorflow//tensorflow/core:framework",
            "@org_tensorflow//tensorflow/core:lib",
            "@com_google_protobuf//:protobuf_headers",
            "@com_google_protobuf//:protobuf",
            ":sentencepiece"
        ]
)

cc_library(
    name = "sentencepiece",
    srcs = [
        "sentencepiece/build/src/libsentencepiece.a",
        "sentencepiece/build/src/libsentencepiece_train.a"
    ],
    hdrs = glob(
        ["sentencepiece/src/*.h"]
    ),
    deps = [
        "@com_google_protobuf//:protobuf_headers",
        "@com_google_protobuf//:protobuf",
    ],
    strip_include_prefix = "sentencepiece/src/"
)
```

4. Run this command to build the docker image. It might takes 10~30 minutes.

   ```sh
   docker build \
       -t tensorflow-serving-with-sentencepiece:local \
       -f sentencepiece.Dockerfile .
   ```

   After finished, you can `docker images` and you will see the image is already there.

   ```sh
   REPOSITORY                              TAG                 IMAGE ID            CREATED             SIZE
   tensorflow-serving-with-sentencepiece   local               37f35ffd1623        21 hours ago        3.78GB
   ```

5. To serve your pretrained model on this image, you can serve like this:

   ```sh
   docker run -d -p $port_pretrain:8501 \
    -v $path_pretrain:$path_pretrain \
    tensorflow-serving-with-sentencepiece:local \
    bash -c "bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server \
    --rest_api_port=8501 --model_name=$name_pretrain --model_base_path=$path_pretrain"
   ```

6. You can verify through curling your serving model:

   ```sh
   curl -d '{"inputs": ["this is a sentence"]}' -X POST http://localhost:8502/v1/models/use-mul-l-1:predict
   ```

Thanks a lot for @stefanondisponibile @sooneecheah your contribution! Without your help, I can't make it serve and run :)

**Comment by hitchhicker** [comment]
We have encountered the same problem caused by **SentencepieceOp**, we have finally succeeded in compiling **tensorflow_text 2.1.0** with the **TF-serving 2.1** release. Our temporary solution is detailed in the issue [tensorflow_text support](https://github.com/tensorflow/serving/issues/1490#issuecomment-571623401).

**Comment by taku910** [comment]
tf-sentencepiece is removed from the repos, as tf.text supports sentencepiece. let me close this bug.
